javascript/ql/src/Security/CWE-078/ShellCommandInjectionFromEnvironment.qhelp

```diff
@@ -0,0 +1,108 @@
+<!DOCTYPE qhelp PUBLIC
+"-//Semmle//qhelp//EN"
+"qhelp.dtd">
+<qhelp>
+<overview>
+	<p>
+
+		Dynamically constructing a shell command with values from the
+		local environment, such as absolute file paths, may inadvertently
+		change the meaning of the shell command. This can result in the shell
+		command misbehaving, or even allow a malicious user to execute
+		arbitrary commands on the system.
+
+	</p>
+
+	<p>
+
+		Such inadvertent changes to the shell command can occur when
+		an environment value contains characters that the shell interprets in a
+		special way, for instance quotes and spaces.
+
+	</p>
+
+
+</overview>
+<recommendation>
+
+	<p>
+
+		If possible, use hard-coded string literals to specify the
+		shell command to run, and provide the dynamic arguments to the shell
+		command separately to avoid interpretation by the shell.
+
+	</p>
+
+	<p>
+
+		Alternatively, if the shell command must be constructed
+		dynamically, then add code to ensure that special characters in
+		environment values does not alter the shell command unexpectedly.
+
+	</p>
+
+</recommendation>
+<example>
+
+	<p>
+
+		The following example shows a dynamically constructed shell
+		command that recursively removes a temporary directory that is located
+		next to the currently executing JavaScript file. Such utilities are
+		often found in ad hoc build systems.
+
+	</p>
+
+	<sample src="examples/shell-command-injection-from-environment.js" />
+
+	<p>
+
+		The shell command will however fail to do as intended if the
+		absolute path of the temporary directory contains spaces. In that
+		case, the shell command will interpret the absolute path as multiple
+		paths, instead of a single path.
+	</p>
+
+	<p>
+
+		For instance, if the absolute path of
+		the temporary directory is <code>/home/username/important
+		project/temp</code>, then the shell command will recursively delete
+		<code>/home/username/important</code> and <code>project/temp</code>,
+		where the latter path gets resolved relative to the working directory
+		of the JavaScript process.
+
+	</p>
+
+
+	<p>
+
+		Even worse, although less likely, a malicious user could
+		control the absolute path of the temporary directory and thereby gain
+		command execution privileges. The shell would then execute command
```

Comment:
I'd suggest turning this sentence around, first showing the malicious path and then explaining what it does.

Reply:
This still needs addressing.

Reply:
Good catch.

Reply:
Fixed.